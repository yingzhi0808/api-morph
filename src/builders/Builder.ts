/**
 * 构建器基础接口，定义所有构建器的通用契约
 *
 * @category Builders
 */
export interface Builder<T> {
  /**
   * 构建最终对象
   * @returns 构建完成的对象
   */
  build(): T;
}
