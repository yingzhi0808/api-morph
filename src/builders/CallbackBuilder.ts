import { cloneDeep } from "radashi";
import type { Builder } from "@/core";
import type { CallbackObject, PathItemObject } from "@/types";

/**
 * 回调构建器，用于构建 OpenAPI CallbackObject
 */
export class CallbackBuilder implements Builder<CallbackObject> {
  private callback: CallbackObject = {};

  build() {
    return cloneDeep(this.callback);
  }

  /**
   * 添加回调表达式。
   * @param expression 运行时计算的表达式，用于标识回调操作的URL。
   * @param pathItem 描述一组请求和预期响应的路径项对象。
   * @returns 回调构建器。
   */
  addExpression(expression: string, pathItem: PathItemObject) {
    const callback = this.callback;
    if (!callback[expression]) callback[expression] = pathItem;
    return this;
  }

  /**
   * 添加扩展字段。
   * @param key 扩展字段键（必须以 'x-' 开头）。
   * @param value 扩展字段值。
   * @returns 回调构建器。
   */
  addExtension(key: `x-${string}`, value: unknown) {
    const document = this.callback;
    if (!document[key]) document[key] = value;
    return this;
  }
}
