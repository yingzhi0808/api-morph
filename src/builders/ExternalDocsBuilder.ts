import { cloneDeep } from "radashi";
import type { Builder } from "@/core";
import type { ExternalDocumentationObject } from "@/types";

/**
 * 外部文档构建器，用于构建 OpenAPI ExternalDocumentationObject
 */
export class ExternalDocsBuilder implements Builder<ExternalDocumentationObject> {
  private externalDocs: ExternalDocumentationObject = { url: "" };

  build() {
    return cloneDeep(this.externalDocs);
  }

  /**
   * 设置外部文档URL。
   * @param url 外部文档URL（必需）。
   * @returns 外部文档构建器。
   */
  setUrl(url: string) {
    this.externalDocs.url = url;
    return this;
  }

  /**
   * 设置外部文档描述。
   * @param description 外部文档描述。
   * @returns 外部文档构建器。
   */
  setDescription(description: string) {
    this.externalDocs.description = description;
    return this;
  }

  /**
   * 添加扩展字段。
   * @param key 扩展字段键（必须以 'x-' 开头）。
   * @param value 扩展字段值。
   * @returns 响应构建器。
   */
  addExtension(key: `x-${string}`, value: unknown) {
    const externalDocs = this.externalDocs;
    if (!externalDocs[key]) externalDocs[key] = value;
    return this;
  }
}
