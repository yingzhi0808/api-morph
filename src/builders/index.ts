/* v8 ignore start */

export * from "./Builder";
export * from "./CallbackBuilder";
export * from "./DocumentBuilder";
export * from "./ExternalDocsBuilder";
export * from "./OperationBuilder";
export * from "./ParameterBuilder";
export * from "./PathItemBuilder";
export * from "./RequestBodyBuilder";
export * from "./ResponseBuilder";
export * from "./SecurityBuilder";
export * from "./ServerBuilder";
