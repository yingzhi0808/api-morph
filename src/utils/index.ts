/* v8 ignore start */

export * from "./string";
export * from "./typeGuards";
