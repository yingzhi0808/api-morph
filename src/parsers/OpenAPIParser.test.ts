import { createProject } from "@tests/utils";
import { describe, expect, it } from "vitest";
import { OpenAPIBuilder } from "@/builders/OpenAPIBuilder";
import { OpenAPIParser } from "./OpenAPIParser";

describe("OpenAPIParser", () => {
  describe("标签解析功能测试", () => {
    it("应该解析包含JSON Schema的JSDoc注释", async () => {
      const project = createProject();
      project.createSourceFile(
        "test.ts",
        `
/**
 * @operation put /api/users/{id}
 * @summary 更新用户信息
 * @description 更新指定用户的个人信息
 * @operationId updateUser
 * @tags users
 * @deprecated
 * @parameter id path required 用户ID
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody required 用户更新数据
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       required: [email]
 *       properties:
 *         email:
 *           type: string
 *           format: email
 *           description: 用户邮箱地址
 *           examples: [john.doe@example.com]
 *         displayName:
 *           type: string
 *           minLength: 2
 *           maxLength: 50
 *           description: 用户显示名称
 *           examples: [John Doe]
 *         avatar:
 *           type: string
 *           format: uri
 *           description: 头像URL
 *           examples: [https://cdn.example.com/avatars/user123.jpg]
 * @response 200 更新用户信息成功
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       properties:
 *         success:
 *           type: boolean
 *           description: 是否成功
 *           examples: [true]
 *         data:
 *           type: object
 *           description: 返回数据
 *           properties:
 *             id:
 *               type: string
 *               format: uuid
 *               description: 用户ID
 *               examples: [123e4567-e89b-12d3-a456-426614174000]
 * @response 404 用户不存在
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       properties:
 *         success:
 *           type: boolean
 *           description: 是否成功
 *           examples: [false]
 *         error:
 *           type: object
 *           properties:
 *             code:
 *               type: string
 *               description: 错误码
 *               examples: [USER_NOT_FOUND]
 *             message:
 *               type: string
 *               description: 错误消息
 *               examples: [指定的用户不存在]
 */
app.put("/api/users/:id", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });

    it("应该解析包含内联schema的JSDoc注释", async () => {
      const project = createProject();
      project.createSourceFile(
        "test.ts",
        `
/**
 * @operation put /api/users/{id}
 * @summary 更新用户信息
 * @description 更新指定用户的个人信息
 * @operationId updateUser
 * @tags users
 * @deprecated
 * @parameter id path required 用户ID
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody {$ref: "#/components/schemas/UpdateUserDto"} required 用户更新数据
 * @response 200 {$ref: "#/components/schemas/UpdateUserVo"} 更新用户信息成功
 * @response 404 {$ref: "#/components/schemas/UserNotFoundVo"} 用户不存在
 */
app.put("/api/users/:id", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder()
        .addSchemaToComponents("UpdateUserDto", {
          type: "object",
          required: ["email"],
          properties: {
            email: {
              type: "string",
              format: "email",
              description: "用户邮箱地址",
              examples: ["john.doe@example.com"],
            },
            displayName: {
              type: "string",
              minLength: 2,
              maxLength: 50,
              description: "用户显示名称",
              examples: ["John Doe"],
            },
            avatar: {
              type: "string",
              format: "uri",
              description: "头像URL",
              examples: ["https://cdn.example.com/avatars/user123.jpg"],
            },
          },
        })
        .addSchemaToComponents("UpdateUserVo", {
          type: "object",
          properties: {
            success: {
              type: "boolean",
              description: "是否成功",
              examples: [true],
            },
            data: {
              type: "object",
              description: "返回数据",
              properties: {
                id: {
                  type: "string",
                  format: "uuid",
                  description: "用户ID",
                  examples: ["123e4567-e89b-12d3-a456-426614174000"],
                },
              },
            },
          },
        })
        .addSchemaToComponents("UserNotFoundVo", {
          type: "object",
          properties: {
            success: {
              type: "boolean",
              description: "是否成功",
              examples: [false],
            },
            error: {
              type: "object",
              properties: {
                code: {
                  type: "string",
                  description: "错误码",
                  examples: ["USER_NOT_FOUND"],
                },
                message: {
                  type: "string",
                  description: "错误消息",
                  examples: ["指定的用户不存在"],
                },
              },
            },
          },
        });

      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });

    it("应该解析包含@link引用的Zod schema的JSDoc注释", async () => {
      const project = createProject({
        tsConfigFilePath: "tsconfig.json",
        useInMemoryFileSystem: false,
        skipAddingFilesFromTsConfig: true,
      });
      project.addDirectoryAtPath("tests/fixtures");
      project.createSourceFile(
        "test.ts",
        `
import { UpdateUserDto, UpdateUserVo, UserNotFoundVo } from "@tests/fixtures/schema"
/**
 * @operation put /api/users/{id}
 * @summary 更新用户信息
 * @description 更新指定用户的个人信息
 * @operationId updateUser
 * @tags users
 * @deprecated
 * @parameter id path required 用户ID
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody {@link UpdateUserDto} required 用户更新数据
 * @response 200 {@link UpdateUserVo} 更新用户信息成功
 * @response 404 {@link UserNotFoundVo} 用户不存在
 */
app.put("/api/users/:id", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });
  });

  describe("Express框架集成测试", () => {
    it("应该能够自动分析Express路由", async () => {
      const project = createProject({
        tsConfigFilePath: "tsconfig.json",
        useInMemoryFileSystem: false,
        skipAddingFilesFromTsConfig: true,
      });
      project.addDirectoryAtPath("tests/fixtures");
      project.createSourceFile(
        "test.ts",
        `
import express from "express"
import { UserVo, UserIdDto, UpdateUserDto, UpdateUserVo, UserNotFoundVo } from "@tests/fixtures/schema"
const app = express()

/**
 * @summary 获取用户信息
 * @description 获取指定用户的个人信息
 * @tags users
 * @response 200 {@link UserVo} 获取用户信息成功
 * @response 404 {@link UserNotFoundVo} 用户不存在
 */
app.get("/api/users/:id", zodValidator({
  params: UserIdDto
}), (req, res) => {})

/**
 * @summary 更新用户信息
 * @description 更新指定用户的个人信息
 * @tags users
 * @response 200 {@link UpdateUserVo} 更新用户信息成功
 * @response 404 {@link UserNotFoundVo} 用户不存在
 */
app.put("/api/users/:id", zodValidator({
  params: UserIdDto,
  body: UpdateUserDto
}), (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });
  });

  describe("选项测试", () => {
    it("应该支持禁用代码分析功能", async () => {
      const project = createProject({
        tsConfigFilePath: "tsconfig.json",
        useInMemoryFileSystem: false,
        skipAddingFilesFromTsConfig: true,
      });
      project.addDirectoryAtPath("tests/fixtures");
      project.createSourceFile(
        "test.ts",
        `
import express from "express"
import { UserVo, UserIdDto, UserNotFoundVo } from "@tests/fixtures/schema"
const app = express()
/**
 * @operation get /api/users/{id}
 * @summary 获取用户信息
 * @description 获取指定用户的个人信息
 * @tags users
 * @response 200 {@link UserVo} 获取用户信息成功
 * @response 404 {@link UserNotFoundVo} 用户不存在
 */
app.get("/api/users/:id", zodValidator({
  params: UserIdDto
}), (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project, { enableCodeAnalysis: false });
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result.paths!["/api/users/{id}"]!.get).toBeDefined();
      expect(result.paths!["/api/users/{id}"]!.get?.parameters).toBeUndefined();
    });
  });
});
