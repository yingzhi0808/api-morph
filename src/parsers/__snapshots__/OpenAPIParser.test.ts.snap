// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OpenAPIParser > Express框架集成测试 > 应该能够自动分析Express路由 1`] = `
{
  "components": {
    "schemas": {
      "UpdateUserDto": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "avatar": {
            "description": "头像URL",
            "examples": [
              "https://cdn.example.com/avatars/user123.jpg",
            ],
            "format": "uri",
            "type": "string",
          },
          "displayName": {
            "description": "用户显示名称",
            "examples": [
              "John Doe",
            ],
            "maxLength": 50,
            "minLength": 2,
            "type": "string",
          },
          "email": {
            "description": "用户邮箱地址",
            "examples": [
              "john.doe@example.com",
            ],
            "format": "email",
            "pattern": "^(?!\\.)(?!.*\\.\\.)([A-Za-z0-9_'+\\-\\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\\-]*\\.)+[A-Za-z]{2,}$",
            "type": "string",
          },
        },
        "required": [
          "email",
        ],
        "type": "object",
      },
      "UpdateUserVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "data": {
            "additionalProperties": false,
            "description": "返回数据",
            "properties": {
              "id": {
                "description": "用户ID",
                "examples": [
                  "123e4567-e89b-12d3-a456-426614174000",
                ],
                "format": "uuid",
                "pattern": "^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000)$",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              true,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "data",
        ],
        "type": "object",
      },
      "UserIdDto": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "id": {
            "description": "用户ID",
            "type": "string",
          },
        },
        "required": [
          "id",
        ],
        "type": "object",
      },
      "UserNotFoundVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "error": {
            "additionalProperties": false,
            "properties": {
              "code": {
                "description": "错误码",
                "examples": [
                  "USER_NOT_FOUND",
                ],
                "type": "string",
              },
              "message": {
                "description": "错误消息",
                "examples": [
                  "指定的用户不存在",
                ],
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              false,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "error",
        ],
        "type": "object",
      },
      "UserVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "avatar": {
            "description": "头像URL",
            "format": "uri",
            "type": "string",
          },
          "createdAt": {
            "description": "创建时间",
            "format": "date-time",
            "pattern": "^(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))T(?:(?:[01]\\d|2[0-3]):[0-5]\\d(?::[0-5]\\d(?:\\.\\d+)?)?(?:Z))$",
            "type": "string",
          },
          "email": {
            "description": "邮箱",
            "format": "email",
            "pattern": "^(?!\\.)(?!.*\\.\\.)([A-Za-z0-9_'+\\-\\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\\-]*\\.)+[A-Za-z]{2,}$",
            "type": "string",
          },
          "id": {
            "description": "用户ID",
            "type": "string",
          },
          "updatedAt": {
            "description": "更新时间",
            "format": "date-time",
            "pattern": "^(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))T(?:(?:[01]\\d|2[0-3]):[0-5]\\d(?::[0-5]\\d(?:\\.\\d+)?)?(?:Z))$",
            "type": "string",
          },
          "username": {
            "description": "用户名",
            "type": "string",
          },
        },
        "required": [
          "id",
          "username",
          "email",
          "createdAt",
          "updatedAt",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/users/{id}": {
      "get": {
        "description": "获取指定用户的个人信息",
        "operationId": "getApiUsersById",
        "parameters": [
          {
            "description": "用户ID",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserVo",
                },
              },
            },
            "description": "获取用户信息成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "获取用户信息",
        "tags": [
          "users",
        ],
      },
      "put": {
        "description": "更新指定用户的个人信息",
        "operationId": "putApiUsersById",
        "parameters": [
          {
            "description": "用户ID",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UpdateUserDto",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UpdateUserVo",
                },
              },
            },
            "description": "更新用户信息成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户信息",
        "tags": [
          "users",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
  ],
}
`;

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含@link引用的Zod schema的JSDoc注释 1`] = `
{
  "components": {
    "schemas": {
      "UpdateUserDto": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "avatar": {
            "description": "头像URL",
            "examples": [
              "https://cdn.example.com/avatars/user123.jpg",
            ],
            "format": "uri",
            "type": "string",
          },
          "displayName": {
            "description": "用户显示名称",
            "examples": [
              "John Doe",
            ],
            "maxLength": 50,
            "minLength": 2,
            "type": "string",
          },
          "email": {
            "description": "用户邮箱地址",
            "examples": [
              "john.doe@example.com",
            ],
            "format": "email",
            "pattern": "^(?!\\.)(?!.*\\.\\.)([A-Za-z0-9_'+\\-\\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\\-]*\\.)+[A-Za-z]{2,}$",
            "type": "string",
          },
        },
        "required": [
          "email",
        ],
        "type": "object",
      },
      "UpdateUserVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "data": {
            "additionalProperties": false,
            "description": "返回数据",
            "properties": {
              "id": {
                "description": "用户ID",
                "examples": [
                  "123e4567-e89b-12d3-a456-426614174000",
                ],
                "format": "uuid",
                "pattern": "^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000)$",
                "type": "string",
              },
            },
            "required": [
              "id",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              true,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "data",
        ],
        "type": "object",
      },
      "UserNotFoundVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "error": {
            "additionalProperties": false,
            "properties": {
              "code": {
                "description": "错误码",
                "examples": [
                  "USER_NOT_FOUND",
                ],
                "type": "string",
              },
              "message": {
                "description": "错误消息",
                "examples": [
                  "指定的用户不存在",
                ],
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              false,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "error",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/users/{id}": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人信息",
        "operationId": "updateUser",
        "parameters": [
          {
            "description": "用户ID",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UpdateUserDto",
              },
            },
          },
          "description": "用户更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UpdateUserVo",
                },
              },
            },
            "description": "更新用户信息成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户信息",
        "tags": [
          "users",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
  ],
}
`;

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含JSON Schema的JSDoc注释 1`] = `
{
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/users/{id}": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人信息",
        "operationId": "updateUser",
        "parameters": [
          {
            "description": "用户ID",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "avatar": {
                    "description": "头像URL",
                    "examples": [
                      "https://cdn.example.com/avatars/user123.jpg",
                    ],
                    "format": "uri",
                    "type": "string",
                  },
                  "displayName": {
                    "description": "用户显示名称",
                    "examples": [
                      "John Doe",
                    ],
                    "maxLength": 50,
                    "minLength": 2,
                    "type": "string",
                  },
                  "email": {
                    "description": "用户邮箱地址",
                    "examples": [
                      "john.doe@example.com",
                    ],
                    "format": "email",
                    "type": "string",
                  },
                },
                "required": [
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "用户更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "返回数据",
                      "properties": {
                        "id": {
                          "description": "用户ID",
                          "examples": [
                            "123e4567-e89b-12d3-a456-426614174000",
                          ],
                          "format": "uuid",
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "success": {
                      "description": "是否成功",
                      "examples": [
                        true,
                      ],
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "更新用户信息成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "error": {
                      "properties": {
                        "code": {
                          "description": "错误码",
                          "examples": [
                            "USER_NOT_FOUND",
                          ],
                          "type": "string",
                        },
                        "message": {
                          "description": "错误消息",
                          "examples": [
                            "指定的用户不存在",
                          ],
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "success": {
                      "description": "是否成功",
                      "examples": [
                        false,
                      ],
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户信息",
        "tags": [
          "users",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
  ],
}
`;

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含内联schema的JSDoc注释 1`] = `
{
  "components": {
    "schemas": {
      "UpdateUserDto": {
        "properties": {
          "avatar": {
            "description": "头像URL",
            "examples": [
              "https://cdn.example.com/avatars/user123.jpg",
            ],
            "format": "uri",
            "type": "string",
          },
          "displayName": {
            "description": "用户显示名称",
            "examples": [
              "John Doe",
            ],
            "maxLength": 50,
            "minLength": 2,
            "type": "string",
          },
          "email": {
            "description": "用户邮箱地址",
            "examples": [
              "john.doe@example.com",
            ],
            "format": "email",
            "type": "string",
          },
        },
        "required": [
          "email",
        ],
        "type": "object",
      },
      "UpdateUserVo": {
        "properties": {
          "data": {
            "description": "返回数据",
            "properties": {
              "id": {
                "description": "用户ID",
                "examples": [
                  "123e4567-e89b-12d3-a456-426614174000",
                ],
                "format": "uuid",
                "type": "string",
              },
            },
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              true,
            ],
            "type": "boolean",
          },
        },
        "type": "object",
      },
      "UserNotFoundVo": {
        "properties": {
          "error": {
            "properties": {
              "code": {
                "description": "错误码",
                "examples": [
                  "USER_NOT_FOUND",
                ],
                "type": "string",
              },
              "message": {
                "description": "错误消息",
                "examples": [
                  "指定的用户不存在",
                ],
                "type": "string",
              },
            },
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              false,
            ],
            "type": "boolean",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/users/{id}": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人信息",
        "operationId": "updateUser",
        "parameters": [
          {
            "description": "用户ID",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UpdateUserDto",
              },
            },
          },
          "description": "用户更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UpdateUserVo",
                },
              },
            },
            "description": "更新用户信息成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户信息",
        "tags": [
          "users",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
  ],
}
`;
